### PR TITLE
Task/654 add apikeyauth to contributions

### PR DESCRIPTION
**Pull request by ywkim312**
## Description
_Please provide a summary of the pull request and the issue it fixes. Please add necessary details, context, dependencies, explanation of when review is needed (see next section), etc._

**Fixes #(add issue number here and remove parentheses)**

[comment]: # (This project only accepts pull requests related to open issues.)
[comment]: # (If suggesting a new feature or change, please discuss it in an issue first.)
[comment]: # (If fixing a bug, there should be an issue describing it with steps to reproduce.)

## Review Time Estimate
_Please give your idea of how soon this pull request needs to be reviewed by selecting one of the options below. This can be based on the criticality of the issue at hand and/or other relevant factors._

[comment]: # (To select an option, please put an 'x' in the applicable box.)
[comment]: # (If you're unsure about any of these, don't hesitate to ask. We're here to help!.)

- [ ] Immediately
- [ ] Within a week
- [x] When possible

## Type of changes
_Please select a relevant option:_

[comment]: # (To select an option, please put an 'x' in the applicable box.)
[comment]: # (If you're unsure about any of these, don't hesitate to ask. We're here to help!.)

- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] Other (any another change that does not fall in one of the above categories.)

## Checklist:
_Please select all applicable options:_

[comment]: # (To select your options, please put an 'x' in the all boxes that apply.)
[comment]: # (If you're unsure about any of these, don't hesitate to ask. We're here to help!.)

- [x] I have signed the Rokwire Contributor License Agreement ([CLA](https://rokwire.org/rokwire_cla)). (_Any contributor who is not an employee of the University of Illinois whose official duties include contributing to the Rokwire software, or who is not paid by the Rokwire project, needs to sign the CLA before their contribution can be accepted._)
- [x] I have updated the [CHANGELOG](../CHANGELOG.md).
- [x] I have read the [Contributor Guidelines](../CONTRIBUTING.md).
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] My change requires updating the documentation.
- [ ] I have made necessary changes to the documentation.
- [ ] I have added tests related to my changes.
- [ ] My changes generate no new warnings.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] Any dependent changes have been merged and published in downstream modules.

---

[comment]: # (Template credit: This pull request template is based on Embedded Artistry {https://github.com/embeddedartistry/templates/blob/master/.github/PULL_REQUEST_TEMPLATE.md}, Clowder {https://github.com/clowder-framework/clowder/blob/develop/.github/PULL_REQUEST_TEMPLATE.md}, and TalAter {https://github.com/TalAter/open-source-templates} templates.)

### CODE REVIEW

**Comment by ywkim312** [comment]
1. The status has been added to contribution building block
2. Most of the endpoints (except POST, DELETE, PUT) has two authentication method (ApiKeyAuth, and GitHubAuth)
3. If ApiKeyAuth, the user will only see published contribution related records
4. If GitHubAuth, the user will see published contribution related records as well as all of the user's record with every status
5. When using ID related endpoint, it will give either 401 or 404. 
6. In ID search, if ApiKey, the user will get 401 when the record with id is not published and the owner is different from the logged in user. 404 for when there is no record with published status
7. In ID search, if GitHub, the user will get 404 if there is no record of published status or if it is not the users record.

**Comment by sandeep-ps** [comment]
> 1. The status has been added to contribution building block
> 2. Most of the endpoints (except POST, DELETE, PUT) has two authentication method (ApiKeyAuth, and GitHubAuth)
> 3. If ApiKeyAuth, the user will only see published contribution related records
> 4. If GitHubAuth, the user will see published contribution related records as well as all of the user's record with every status
> 5. When using ID related endpoint, it will give either 401 or 404.
> 6. In ID search, if ApiKey, the user will get 401 when the record with id is not published and the owner is different from the logged in user. 404 for when there is no record with published status
> 7. In ID search, if GitHub, the user will get 404 if there is no record of published status or if it is not the users record.

@ywkim312, in item 6, when ApiKey is provided for say GET /contributions/{id} endpoint, if it's an unpublished contribution, the building block should return 401, right? I'm getting 404 for the below example contribution when using ApiKey:

```
{
    "capabilities": [
        {
            "apiBaseUrl": null,
            "apiDocUrl": null,
            "contacts": null,
            "dataDeletionEndpointDetails": {
                "apiKey": "apiKey",
                "deletionEndpoint": "deletionEndpoint",
                "description": "Laboriosam libero itaque impedit veniam doloremque in voluptates dolorum natus."
            },
            "deploymentDetails": {
                "authMethod": null,
                "databaseDetails": null,
                "dockerImageName": null,
                "environmentVariables": null,
                "location": "internal"
            },
            "description": "Illo est quasi id quia cumque in eligendi architecto.",
            "healthCheckUrl": "healthCheckUrl",
            "isOpenSource": true,
            "name": "Gorgeous Fresh Keyboard 789",
            "status": "Submitted",
            "version": "version"
        }
    ],
    "contributionAdmins": [
        "sandeep-ps"
    ],
    "contributors": [
        {
            "address": null,
            "email": null,
            "name": null,
            "phone": null
        }
    ],
    "dateCreated": "2021/04/19T13:09:51",
    "dateModified": "2021/04/19T13:09:51",
    "longDescription": null,
    "name": "Intelligent Metal Cheese 410",
    "shortDescription": "Nam necessitatibus accusamus ut excepturi.",
    "talents": [
        {
            "dataDescription": null,
            "endDate": null,
            "longDescription": null,
            "minEndUserRoles": null,
            "minUserPrivacyLevel": null,
            "name": "Awesome Fresh Sausages 234",
            "requiredBuildingBlocks": null,
            "requiredCapabilities": null,
            "selfCertification": {
                "dataDeletionUponRequest": null,
                "discloseAds": null,
                "discloseImageRights": null,
                "discloseSponsors": null,
                "respectingUserPrivacySetting": null
            },
            "shortDescription": "Qui est laboriosam asperiores beatae perspiciatis repellat impedit et odit.",
            "startDate": null
        }
    ]
}
```

File: contributions/api/controllers/contributions.py [hunk 1]
```diff
@@ -147,11 +147,14 @@ def post(token_info):
...
+def search(token_info=None, name=None):
     # args = request.args
-
-    query = dict()
     is_list = False
+    query = dict()
+
+    login_id, is_login = get_login(token_info)
+
+    query = query_params.format_query_status_login(query, login_id, is_login)
```
[the remark this answers is not in the document]
Comment (ywkim312) [reply contributions/api/controllers/contributions.py:157]: Yes, I have already developed code for this in my current working branch.

[the remark this answers is not in the document]
Comment (sandeep-ps) [reply contributions/api/controllers/contributions.py:157]: Okay. Thanks.

**Comment by ywkim312** [comment]
> @ywkim312, in item 6, when ApiKey is provided for say GET /contributions/{id} endpoint, if it's an unpublished contribution, the building block should return 401, right? I'm getting 404 for the below example contribution when using ApiKey:

Sorry, the number 6 is wrong because there is no way to find out who is the user of the endpoint when using ApiKey. It will only get 404 if the it is using ApiKey.

**Comment by ywkim312** [comment]
> > @ywkim312, in item 6, when ApiKey is provided for say GET /contributions/{id} endpoint, if it's an unpublished contribution, the building block should return 401, right? I'm getting 404 for the below example contribution when using ApiKey:
> 
> Sorry, the number 6 is wrong. It will get 404 when there is no record, and 401 when there is no published record. I will update nubmer 6's description and code

**Comment by ywkim312** [comment]
There is an issue in my branch. I will update the fixes in my current working branch... Sorry about that.

**Comment by sandeep-ps** [comment]
Please fix the merge conflicts. The tests so far look fine.

**Comment by ywkim312** [comment]
> Please fix the merge conflicts. The tests so far look fine.

fixed

**Review by sandeep-ps — APPROVED** [review]

**Comment by sandeep-ps** [comment]
So far, all the tests are successful except for the status field. Merging this PR. We can open any remaining related issues separately. Thanks.

**Pull request merged**